### PR TITLE
Enable changing passwords in user_ldap

### PR DESCRIPTION
**Pull request by GitHubUser4234**
Continuing PR #600 here due to some GitHub issues, sorry for inconvenience caused. 😅 

@blizzz : All your comments have hopefully been taken care of :)

### CODE REVIEW

File: apps/user_ldap/lib/LDAP.php [hunk 1]
```diff
@@ -194,6 +195,16 @@ public function search($link, $baseDN, $filter, $attr, $attrsOnly = 0, $limit =
 
 	/**
 	 * @param LDAP $link
+* @param string $userDN
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/LDAP.php]: Some whitespace is missing here

Comment (GitHubUser4234) [inline apps/user_ldap/lib/LDAP.php]: Should be fine now, thanks ~

**Comment by GitHubUser4234** [comment]
> image

Ok, this has been discussed in IRC and @blizzz had this great idea to make the error message a combination of a generic translatable message plus a hint message containing the non-translatable message returned by the LDAP server. 

I will apply this tmw.

**Comment by GitHubUser4234** [comment]
> > image
> 
> Ok, this has been discussed in IRC and @blizzz had this great idea to make the error message a combination of a generic translatable message plus a hint message containing the non-translatable message returned by the LDAP server. 
> 
> I will apply this tmw.

Applied & pushed.

File: apps/user_ldap/lib/LDAP.php [hunk 2]
```diff
@@ -194,6 +197,20 @@ public function search($link, $baseDN, $filter, $attr, $attrsOnly = 0, $limit =
 
 	/**
 	 * @param LDAP $link
+	 * @param string $userDN
+	 * @param string $password
+	 * @return bool
+	 */
+	public function modReplace($link, $userDN, $password) {
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAP.php]: Could you also add this method to the ILDAPWrapper interface, please?

File: apps/user_ldap/lib/LDAP.php [hunk 3]
```diff
@@ -194,6 +197,20 @@ public function search($link, $baseDN, $filter, $attr, $attrsOnly = 0, $limit =
...
+		try {
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAP.php]: Could we move the try-and-catch logic to Access::setPassword() to avoid logic in here? Would also avoid  "funny" messages if it was used for another reason in the future while the same base error is thrown.

File: apps/user_ldap/lib/LDAP.php [hunk 4]
```diff
@@ -194,6 +197,20 @@ public function search($link, $baseDN, $filter, $attr, $attrsOnly = 0, $limit =
...
+			return $this->invokeLDAPMethod('mod_replace', $link, $userDN, array('userPassword' => $password));
+		} catch(ConstraintViolationException $e) {
+			throw new HintException('Password change rejected. Hint: '.$e->getMessage(), '', $e->getCode());
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAP.php]: Probably I was not clear about the second parameter which is currently the emtpy string, sorry. That one is the "Hint" parameter which accepts also text. I would pass ``$e->getMessage()`` to it. As first parameter ``'Password change rejected.'`` would be good. To tranlsate that we would need to get from ``\OC::$server->getL10N('user_ldap');``, as we do not inject it at this point (in neither Access not User_LDAP). Injecting it  is cumbersome of course with the necessary changes, so for I10L it's probably survivable to not do it for now…

File: apps/user_ldap/templates/settings.php [hunk 5]
```diff
@@ -101,6 +101,8 @@
...
+				<p><label for="ldap_turn_on_pwd_change"><?php p($l->t('Enable LDAP password changes per user'));?></label><span class="inlinetable"><span class="tablerow left"><input type="checkbox" id="ldap_turn_on_pwd_change" name="ldap_turn_on_pwd_change" value="1" data-default="<?php p($_['ldap_turn_on_pwd_change_default']); ?>" title="<?php p($l->t('Allow LDAP users to change their password and allow Super Administrators and Group Administrators to change the password of their LDAP users. Only works when access control policies are configured accordingly on the LDAP server. As passwords are sent in plaintext to the LDAP server, transport encryption must be used and password hashing should be configured on the LDAP server.'));?>" /><span class="tablecell">(New password is sent as plain text to LDAP)</span></span>
```
Comment (blizzz) [inline apps/user_ldap/templates/settings.php]: Wrap the new "(New password…" text into ``<?php p($l->t($text)); ?>`` so it can get translated

File: apps/user_ldap/lib/Access.php [hunk 6]
```diff
@@ -221,6 +221,26 @@ public function readAttribute($dn, $attr, $filter = 'objectClass=*') {
 		\OCP\Util::writeLog('user_ldap', 'Requested attribute '.$attr.' not found for '.$dn, \OCP\Util::DEBUG);
 		return false;
 	}
+	
+	/**
+	 * Set password for an LDAP user identified by a DN
+	 * @param string $userDN the user in question
+	 * @param LDAP $password the new password
```
Comment (blizzz) [inline apps/user_ldap/lib/Access.php]: $password should be of type string

File: apps/user_ldap/lib/User_LDAP.php [hunk 7]
```diff
@@ -175,6 +176,28 @@ public function checkPassword($uid, $password) {
 	}
 
 	/**
+	 * Set password
+	 * @param string $uid The username
+	 * @param string $password The new password
+	 * @return bool
+	 */
+	public function setPassword($uid, $password) {
+		$ldapRecord = $this->getLDAPUserByLoginName($uid);
```
Comment (blizzz) [inline apps/user_ldap/lib/User_LDAP.php]: ``$uid`` is not the login name, but the internal username. If both differ, it won't work (seen here while testing). You can directly fetch the corresponding user by ``$user = $this->access->userManager->get($uid);``. Then it'll work :)

**Review by blizzz — CHANGES_REQUESTED** [review]

**Comment by blizzz** [comment]
Btw, with the next commit please do a signoff like ``git commit --signoff``, (see Sign Your Work on https://github.com/nextcloud/server/blob/master/CONTRIBUTING.md)

File: apps/user_ldap/lib/LDAP.php (hunk 2 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/LDAP.php]: ok~

File: apps/user_ldap/lib/LDAP.php (hunk 3 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/LDAP.php]: ok~

File: apps/user_ldap/lib/LDAP.php (hunk 4 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/LDAP.php]: I tried, but the result was that the first parameter didn't get displayed in the UI. When looking at the HintException code it looks like it either has to be first or second parameter?

```
public function getHint() {
	if (empty($this->hint)) {
		return $this->message;
	}
	return $this->hint;
}
```

File: apps/user_ldap/templates/settings.php (hunk 5 above)
Comment (GitHubUser4234) [inline apps/user_ldap/templates/settings.php]: oops, fixed that one 😅

File: apps/user_ldap/lib/Access.php (hunk 6 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/Access.php]: yep~

File: apps/user_ldap/lib/User_LDAP.php (hunk 7 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/User_LDAP.php]: Ok~~ I saw something similar in the` checkPassword()` method, does this apply there, too?

Comment (GitHubUser4234) [inline apps/user_ldap/lib/User_LDAP.php]: Oh i see, then the naming of that parameter doesn't seem to reflect it, at least not for an outsider like me 😆  If acceptable it might be helpful in the future when we'd change it from `$uid` to `$loginName`? I saw that some classes like `OC\User\Manager` are in fact doing that while in fact most classes don't.

Comment (blizzz) [inline apps/user_ldap/lib/User_LDAP.php]: Totally agree!

File: apps/user_ldap/lib/LDAP.php (hunk 4 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/LDAP.php]: As discussed in IRC, necessary changes in core will be conducted, so that the UI can show both parameters.

File: apps/user_ldap/lib/User_LDAP.php (hunk 7 above)
Comment (GitHubUser4234) [inline apps/user_ldap/lib/User_LDAP.php]: As discussed in IRC, the name change is fine, but would be done in a separate PR for fixing variable names.

**Comment by GitHubUser4234** [comment]
After discussion in #1787 this got updated, rebased and pushed.

**Comment by GitHubUser4234** [comment]
@blizzz : great thing is the personal page bug #1874 has been fixed, so this can be tested without obstacles.

**Review by blizzz — APPROVED** [review]

**Comment by blizzz** [comment]
Tested and works 👍 

On the Users page, when changing a password and confirming with pressing Enter, the response handling fails (no confirmation, input field stays with focus, despite successful change - but some in error case), but this issue is also present with master and thus out of scope. Password change on personal page works.

Another reviewer please? @nextcloud/ldap @LukasReschke

File: apps/user_ldap/lib/Access.php [hunk 8]
```diff
@@ -221,6 +223,30 @@ public function readAttribute($dn, $attr, $filter = 'objectClass=*') {
...
 		return false;
 	}
+	
+	/**
+	 * Set password for an LDAP user identified by a DN
+	 * @param string $userDN the user in question
+	 * @param string $password the new password
+	 * @return bool
+	 */
+	public function setPassword($userDN, $password) {
```
Comment (LukasReschke) [inline apps/user_ldap/lib/Access.php]: Missing tests. Will add myself so we get this done.

File: apps/user_ldap/lib/User_LDAP.php [hunk 9]
```diff
@@ -175,6 +176,26 @@ public function checkPassword($uid, $password) {
...
+	 * Set password
+	 * @param string $uid The username
+	 * @param string $password The new password
+	 * @return bool
+	 */
+	public function setPassword($uid, $password) {
+		$user = $this->access->userManager->get($uid);
+
+		if(!$user instanceof User) {
+			throw new \Exception('LDAP setPassword: Could not get user object for uid ' . $uid .
```
Comment (LukasReschke) [inline apps/user_ldap/lib/User_LDAP.php]: Not covered by tests. Will add.

File: apps/user_ldap/lib/User_LDAP.php [hunk 10]
```diff
@@ -175,6 +176,26 @@ public function checkPassword($uid, $password) {
...
+				'. Maybe the LDAP entry has no set display name attribute?');
+		}
+		if($user->getUsername() !== false) {
+			return $this->access->setPassword($user->getDN(), $password);
+		}
+
+		return false;
```
Comment (LukasReschke) [inline apps/user_ldap/lib/User_LDAP.php]: Not covered by tests. Will add.

File: apps/user_ldap/lib/User_Proxy.php [hunk 11]
```diff
@@ -262,6 +262,17 @@ public function getDisplayNames($search = '', $limit = null, $offset = null) {
...
+	
+	/**
+	 * Set password
+	 * @param string $uid The username
+	 * @param string $password The new password
+	 * @return bool
+	 *
+	 */
+	public function setPassword($uid, $password) {
+		return $this->handleRequest($uid, 'setPassword', array($uid, $password));
```
Comment (LukasReschke) [inline apps/user_ldap/lib/User_Proxy.php]: Not covered by tests. Will add.

**Review by LukasReschke — CHANGES_REQUESTED** [review]

**Comment by LukasReschke** [comment]
Replaced by https://github.com/nextcloud/server/pull/2286 where I added some tests.

**Comment by LukasReschke** [comment]
And merged 🚀

**Comment by GitHubUser4234** [comment]
Wow, finally the rocket took off and that at light speed! Thanks A LOT @blizzz, @LukasReschke, @MorrisJobke for the effort put into this ✨
